src/order/countable_dense_linear_order.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2020 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import order.ideal
+import data.finset
+
+/-!
+# The back and forth method and countable dense linear orders
+
+## Results
+
+Suppose `α β` are linear orders, with `α` countable and `β` dense, nonempty, without endpoints.
+Then there is an order embedding `α ↪ β`. If in addition `α` is dense, nonempty, without
+endpoints and `β` is countable, then we can upgrade this to an order isomorhpism `α ≃ β`.
+
+The idea for both results is to consider "partial isomorphisms", which
+identify a finite subset of `α` with a finite subset of `β`, and prove that
+for any such partial isomorphism `f` and `a : α`, we can extend `f` to
+include `a` in its domain.
+
+## References
+
+https://en.wikipedia.org/wiki/Back-and-forth_method
+
+## Tags
+
+back and forth, dense, countable, order
+
+-/
+
+noncomputable theory
+open_locale classical
+open order
+
```

Comment:
Question (also to other reviewers): should (some of) these declarations be namespaced?

Reply:
Probably. The declarations are now all in the `order` namespace, and the auxiliary declarations are further down in `order.partial_iso`.